central/auth/service/service_impl.go

```diff
@@ -89,3 +126,108 @@ func authStatusForID(id authn.Identity) (*v1.AuthStatus, error) {
+	}
+	return &v1.GetAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(config)}, nil
+}
+
+func (s *serviceImpl) AddAuthMachineToMachineConfig(ctx context.Context, request *v1.AddAuthMachineToMachineConfigRequest) (*v1.AddAuthMachineToMachineConfigResponse, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(config, true); err != nil {
+		return nil, err
+	}
+	config.Id = uuid.NewV4().String()
+	storageConfig, err := s.authDataStore.AddAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.AddAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
+}
+
+func (s *serviceImpl) UpdateAuthMachineToMachineConfig(ctx context.Context, request *v1.UpdateAuthMachineToMachineConfigRequest) (*v1.Empty, error) {
+	config := request.GetConfig()
+	if err := s.validateAuthMachineToMachineConfig(config, false); err != nil {
+		return nil, err
+	}
+
+	if err := s.authDataStore.UpdateAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config)); err != nil {
+		return nil, err
+	}
+
+	return &v1.Empty{}, nil
+}
+
+func (s *serviceImpl) DeleteAuthMachineToMachineConfig(ctx context.Context, id *v1.ResourceByID) (*v1.Empty, error) {
+	if err := s.authDataStore.RemoveAuthM2MConfig(ctx, id.GetId()); err != nil {
+		return nil, errox.InvalidArgs.
+			Newf("could not delete auth machine to machine config with id %q", id.GetId()).CausedBy(err)
+	}
+	return &v1.Empty{}, nil
+}
+
+func (s *serviceImpl) ExchangeAuthMachineToMachineToken(_ context.Context, _ *v1.ExchangeAuthMachineToMachineTokenRequest) (*v1.ExchangeAuthMachineToMachineTokenResponse, error) {
+	return nil, errox.InvariantViolation.New("not yet implemented")
+}
+
+func (s *serviceImpl) validateAuthMachineToMachineConfig(config *v1.AuthMachineToMachineConfig, skipIDCheck bool) error {
+	if config == nil {
+		return errox.InvalidArgs.New("empty config given")
+	}
+	if config.GetId() == "" && !skipIDCheck {
+		return errox.InvalidArgs.New("empty ID given")
+	}
+
+	duration, err := time.ParseDuration(config.GetTokenExpirationDuration())
+	if err != nil {
+		return fmt.Errorf("%w: %w: %w", errox.InvalidArgs, errInvalidToken, err)
+	}
+
+	if duration < time.Minute || duration > 24*time.Hour {
+		return errox.InvalidArgs.Newf("token expiration must be between 1 minute and 24 hours, but was %s",
+			duration.String())
+	}
+
+	if config.GetType() == v1.AuthMachineToMachineConfig_GENERIC && config.GetIssuer() == "" {
+		return errox.InvalidArgs.Newf("type %s was used, but no configuration for the issuer was given",
+			storage.AuthMachineToMachineConfig_GENERIC)
+	}
+
+	if config.GetIssuer() != "" {
+		_, err := url.Parse(config.GetIssuer())
+		if err != nil {
+			return fmt.Errorf("%w: %w: %w", errox.InvalidArgs, errInvalidIssuer, err)
+		}
+		config.Issuer = urlfmt.FormatURL(config.GetIssuer(), urlfmt.HTTPS, urlfmt.NoTrailingSlash)
```

Comment:
Given what the RFC <https://datatracker.ietf.org/doc/html/rfc7519#section-2> says about <StringOrURI>, we must not even try to format the URL if it includes `:` the colon character.

The RFC also says we must not canonicalize the URL, but I'm torn about it. There will, certainly, be user misconfiguration errors. Perhaps these should be left as-is, and the first attempt to exchange the token would fail indicating the error.
However, I'm not sure about cases when the config was first set correctly but the issuer in actual tokens changes after some time in an "insignificant detail" such as the trailing slash.
WDYT?

Reply:
I think you're looking at the wrong RFC.

While on the generic JWT RFC the issuer may not be a URL,[ for the OIDC RFC](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) (which is what we are actually using, since this is expected to be an OIDC issued ID token) it is required to be a URL with the `https` scheme.

I only kept the URL parsing + HTTPS scheme validation in the function, and haven't changed the issuer itself. I think we can rely on the token exchange to fail to realize potential issues with the issuer (i.e. misconfigurations) which is hard to tell at this point from our validation logic.